### PR TITLE
Avoid using FutureNCCL before it's ready

### PR DESCRIPTION
**Pull request by lw**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48790 Add support for async callbacks in ivalue::Future
* #49014 Drop FutureNCCL in favor of vanilla CUDAFuture
* #48789 Make CUDAFuture remember and restore current device in callback
* #48840 Remove DataPtr extractor from CUDAFuture
* #48788 Cache the DataPtrs in CUDAFuture
* #48506 Split out reusable CUDAFuture from FutureNCCL
* #48505 Merge common parts of FutureNCCL into at::ivalue::Future
* #48504 Split FutureNCCL's CUDA-specific parts from generic future logic
* #48502 Support wider range of types in FutureNCCL
* #48501 Don't store device indices separately on FutureNCCL
* #48500 Add multi-GPU support to FutureNCCL
* #48563 Fix FutureNCCL not recording dataptrs with caching alloc in wait()
* #48503 Fix FutureNCCL's completed() disagreeing with wait()
* #48499 Record CUDA events for "follow-up" FutureNCCL inside markCompleted
* #48498 Use fresh stream from pool for each FutureNCCL callback
* #48497 Make FutureNCCL record events in current stream
* #48496 Have FutureNCCL record streams w/ allocator in addCallback
* #48562 Add some safeguards to FutureNCCL
* #48495 Remove NCCL dependency from PythonFutureWrapper
* **#48561 Avoid using FutureNCCL before it's ready**

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

WorkNCCL allows to extract a FutureNCCL through getFuture(). There is one instance of this method being called by ProcessGroupNCCL itself, in order to attach a callback to it. This was happening _before_ the work was actually launched, however FutureNCCL does _always_ invoke its callbacks immediately inline. The events that the FutureNCCL was using hadn't been recorded yet, thus blocking on them was a no-op. Moreover, the function that was being called was installed by the generic ProcessGroup superclass, which is not CUDA-aware, and thus probably didn't make any use of the CUDA events or streams.

https://github.com/pytorch/pytorch/blob/383abf1f0c1f74e0f471d47e505895d1b0e6bb20/torch/lib/c10d/ProcessGroup.cpp#L66

In short: I believe that creating a FutureNCCL and attaching a callback was equivalent to just invoking that function directly, without any CUDA-specific thing. I'm thus converting the code to do just that, in order to simplify it.

Note that, given the comment, I don't think this was the original intention of that code. It seems that the function was intended to be run once the work finished. However, I am not familiar with this code, and I don't want to introduce any functional changes.

Differential Revision: [D25210337](https://our.internmc.facebook.com/intern/diff/D25210337/)

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -1073,15 +1073,13 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
...
   if (work->recordFunctionEndCallback_) {
     // recordFunctionEndCallback_ is normally called in fininsh() function by
-    // base class, but since finish is not called by WorkNCCL, we schedule this
-    // function to be run when work is done.
+    // base class, but since finish is not called by WorkNCCL, we run this
+    // function now.
     // Note when can_profile is false, profilingTitle is not provided and so,
     // recordFunctionEndCallback_ is not set.
-    work->getFuture()->addCallback(std::move(work->recordFunctionEndCallback_));
+    work->recordFunctionEndCallback_();
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:1080]: > The events that the FutureNCCL was using hadn't been recorded yet, thus blocking on them was a no-op. 

Does this mean the existing code for launching the cb was wrong?

**Review by mrshenli — APPROVED** [review]
This LGTM as it does not change the existing behavior.

cc @SciPioneer (for FutureNCCL) @rohan-varma (for profiler) please comment if there are concerns. The question I have is that does this mean it does not correctly profile the execution time of NCCL c10d operations?

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (lw) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1080]: The code of FutureNCCL was correct (because, for NCCL, it's "correct" to always invoke callbacks inline).

The mistake was in ProcessGroupNCCL, which was constructing a FutureNCCL with "incomplete" arguments, but then using the FutureNCCL before making those arguments complete. (These "arguments" are the CUDA events).

Once FutureNCCL was returned by ProcessGroupNCCL, its arguments were complete, so the users of ProcessGroupNCCL couldn't hit this issue.

Comment (lw) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1080]: I still believe that the high-level behavior here is wrong, since it caused recordFunctionEndCallback_ to be invoked _before_ the NCCL function was called. But this commit does nothing more than make that mistake explicit. I think it should be fixed in a separate PR by someone who knows what's going on.

Comment (rohan-varma) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1080]: Yes, this is a bug and the reason why we disabled some of the failing tests as part of https://github.com/pytorch/pytorch/pull/48129. Attempting to fix/debug this as part of https://github.com/pytorch/pytorch/pull/48664, but in the meantime this PR should be fine as-is.

[the remark this answers is not in the document]
Comment (mrshenli) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1080]: > since the callback is CPU only when the profiler is enabled without `use_cuda=True`.

Does this mean ideally we need to record the TLS including `use_cuda=True` when calling `addCallback` (if it can behave as normal future.addCallback). And because `FutureNCCL` always run cb inline, we don't even have the opportunity to check the state for `use_cuda`?

Question, for c10d ops, when/where does profiler initialization/enabling happen?

[the remark this answers is not in the document]
Comment (lw) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1080]: I don't know how the profiler is designed to work with CUDA, but it does sound odd to me to try to measure GPU timing using CPU code. I though the proper tool for the job were CUDA events, created with the EnableTiming flag. If that can work, then such events can be created and enqueued even from an inline CPU callback, and CUDA should still be able to correctly collect timing for them.

[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1080]: @lw Right, to profile GPU code the user should invoke the profiler with use_cuda=True.

When this happens, this code path is run, which indeed uses CUDA events to do the timing: 
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/profiler_cuda.cpp#L35

@mrshenli, I think with NCCL we don’t have to propagate any TLS State since the RecordFunction start and end happen in the same thread synchronously. Even so, with the current way we fork the ProfilerState, the use_cuda flag is carried along so events on another thread would be profiled with cuda.

[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1080]: @mrshenli For profiler init, that would be by the user invoking python profiler with c10d ops wrapped inside. For RecordFunction initialization, basically RecordFunction::start, it happens here: https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroup.cpp#L60
